### ISSUE TITLE
"flutter run -d chrome" dev server responds with 200 status/homepage HTML for missing files

### ISSUE THREAD

**Issue by DanTup** [flutter/flutter#67053]
Using current master:

- flutter create foo
- cd foo
- flutter run -d chrome
- when the browser opens, delete the `#` from the URL and add `/missing`

The response returns a 200 status with HTML for the app, but I'd expect a 404.

![Screenshot 2020-10-01 at 15 39 35](https://user-images.githubusercontent.com/1078012/94823919-50f7ef80-03fc-11eb-9b44-fde2de6f69a5.png)

**Comment by jonahwilliams** [comment]
@mdebbar

**Comment by Vitor-Bukovitz** [comment]
Same here but I'm getting my website regardless of what comes after `/`, e.g: `https://www.mywebsite.com/12398ueushadihsa12iu`
After it loads the url becomes like thus: `https://www.mywebsite.com/12398ueushadihsa12iu#/`

**Comment by mdebbar** [comment]
This is WAI. The reason we made it this way is to prepare to support regular paths in flutter web apps (paths without the `#`). So if you configure your app to use `PathUrlStrategy` (instead of the old `HashUrlStrategy`) you can visit: `http://localhost:8080/profile` which loads your app, and use `/profile` as a route name for the Navigator.

That said, I would like to learn more if this is causing any issues for you.

**Comment by DanTup** [comment]
It caused an issue in DevTools, as it was trying to check whether it was running through its own server (so sending a request to the server to `api/ping` and checking for a 200 status). That's not a significant issue though and it was worked around by checking the contentType was not text/html (https://github.com/flutter/devtools/pull/2396). I don't think if this kind of thing is common.

I hadn't considered `PathUrlStrategy` - I think supporting that is more valuable than avoiding the DevTools workaround though (and we could work around it other says, like inspecting the response from the API is a known good value) so I wouldn't change this if that's the only issue.

Thanks!

**Comment by DanTup** [comment]
@mdebbar out of interest, would serving a 404 status but still the page content work? It might work better if people are sending API calls and checking status codes, and I don't *think* it'd make any difference to the users refreshing the pages in their browser?

**Comment by mdebbar** [comment]
@DanTup serving 404 tells the browser (and the user) that the content they are requesting doesn't exist. We can still serve some content with 404, but that content is supposed to be the 404 page.

Would it be possible to serve your flutter app on `http://localhost:12345/devtools/` (instead of `http://localhost:12345/`)? You can do that by adding `<base href="/devtools/">` to the `web/index.html` file. If you do this, all requests that aren't prefixed with `/devtools/` will get a 404 response. So your `/api/ping` request will receive 404 response as you expect.

Let me know if this workaround is okay for you.

**Comment by DanTup** [comment]
@mdebbar I made [another workaround](https://github.com/flutter/devtools/pull/2411) which inspects the response to api/ping is the content we sent which is probably better. The 404 was really an idea in case other people are doing the same sort of thing.

Though out of interest, how would adding a `base href` change the response for api/ping? Is `flutter run` reading the `base href` and behaving differently?

**Comment by mdebbar** [comment]
@DanTup yes, `flutter run` reads `base href` and only serves requests under that path. So if base href is `/foo/bar/` and you send a request to `/api/ping/` the server returns a 404.

**Comment by DanTup** [comment]
Ah, good to know! I think the current workaround is probably simpler for now, but that might be an idea if we end up with other issues (or need more than just the api/ping to work that way). Thanks!

**Comment by yjbanov** [comment]
Reclassifying this as a feature request to support matching on path patterns that are valid route paths and only serving `index.html` to them. We already do some basic filtering of HTTP requests, but giving the developer more control would help even more.

**Comment by kevmoo** [comment]
Here's my vote here. I honestly think we should let folks opt-in to the "serve index file everywhere else" model. 404s are USEFUL!
